Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,246 @@
+    /// The configuration object of the producer client.
+    private var config: KafkaConfig
+    /// The ``KafkaTopicConfig`` used for newly created topics.
     private let topicConfig: KafkaTopicConfig
+    /// A logger.
+    private let logger: Logger
+    /// Dictionary containing all topic names with their respective `rd_kafka_topic_t` pointer.
+    private var topicHandles: [String: OpaquePointer]
 
-    // Preliminary implementation
+    // Implicitly unwrapped optionals because we must pass self to our config before initializing client etc.
+    /// Unmanaged reference to self with an unbalanced retain.
+    private var referenceToSelf: UnsafeMutableRawPointer!
+    /// Used for handling the connection to the Kafka cluster.
+    private var client: KafkaClient!
+    /// Refrence to `librdkafka`'s `rd_kafka_t` instance.
+    private var kafkaHandle: OpaquePointer!
+    /// Task that polls the Kafka cluster for updates periodically.
+    private var pollTask: Task<Void, Never>!
+
+    /// Initialize a new `KafkaProducer`.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the `KafkaProducer`.
+    /// - Parameter topicConfig: The ``KafkaTopicConfig`` used for newly created topics.
+    /// - Parameter logger: A logger.
     public init(
         config: KafkaConfig = KafkaConfig(),
         topicConfig: KafkaTopicConfig = KafkaTopicConfig(),
         logger: Logger
-    ) throws {
-        self.client = try KafkaClient(type: .producer, config: config, logger: logger)
+    ) async throws {
+        self.config = config
         self.topicConfig = topicConfig
+        self.logger = logger
+        self.topicHandles = [:]
+        self.state = .started
+
+        // Create a new reference to self that is then passed
+        // to the KafkaConfig as an opaque pointer.
+        // To avoid memory leaks, this reference is evantually
+        // removed in the shutdownGracefully method.
+        self.referenceToSelf = Unmanaged.passRetained(self).toOpaque()
+
+        self.config.setDeliveryReportCallback(callback: self.deliveryReportCallback)
+        self.config.setOpaque(opaque: self.referenceToSelf)
+
+        self.client = try KafkaClient(type: .producer, config: self.config, logger: self.logger)
+        self.kafkaHandle = self.client.kafkaHandle
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [kafkaHandle] in
+            while !Task.isCancelled {
+                rd_kafka_poll(kafkaHandle, 0)
+                try? await Task.sleep(nanoseconds: 1_000_000)
+            }
+        }
+    }
+
+    /// Method to shutdown the ``KafkaProducer``.
+    ///
+    /// This method flushes any buffered messages and waits until a callback is received for all of them.
+    /// Afterwards, it shuts down the connection to Kafka and cleans any remaining state up.
+    public func shutdownGracefully() throws {
+        try failIfProducerShutDown()
+        self.state = .shuttingDown
```

Comment:
Instead of doing this can you please explicitly `switch` over the state here. This is a common pattern in the swift on server space to make sure we are only doing state transition at the right time.

I would recommend to have a `switch` over the state in this method and then a separate `private` method called `_ shutdownGracefully` that does the actual shutdown and only call this in the case of `.started`